### PR TITLE
Adding MoE recipe for external aerodynamics

### PR DESCRIPTION
**Pull request by mnabian**
# PhysicsNeMo Pull Request

Adding MoE recipe for external aerodynamics

## Checklist

- [x] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/physicsnemo/blob/main/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.
- [x] The [CHANGELOG.md](https://github.com/NVIDIA/physicsnemo/blob/main/CHANGELOG.md) is up to date with these changes.
- [x] An [issue](https://github.com/NVIDIA/physicsnemo/issues) is linked to this pull request.

### CODE REVIEW

**Review by Dibyajyoti-Chakraborty — COMMENTED** [review]
Looks Fine.

File: examples/cfd/external_aerodynamics/moe/conf/config.yaml [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+use_moe_bias: false                                        # Whether to use bias in the MoE output layer
+include_normals: true                                      # Whether to include surface normals as additional input features
+checkpoint_dir: checkpoints                                # Directory to save/load model checkpoints
+
+# ┌───────────────────────────────────────────┐
+# │          Training Configuration           │
+# └───────────────────────────────────────────┘
+
+lambda_entropy: 0.01                                       # Weight for entropy regularization (encourages expert diversity)
+batch_size: 1                                              # Batch size for training (code is only tested with batch size 1)
```
Comment (Dibyajyoti-Chakraborty) [inline examples/cfd/external_aerodynamics/moe/conf/config.yaml:66]: The batch size = 1 part is a bit concerning (given that in readme the batch size is 32). Could you give me some explanation in comments?

File: examples/cfd/external_aerodynamics/moe/preprocessor.py [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+
+    pMeanTrim_mean = np.mean(pMeanTrim_all)
+    pMeanTrim_std = np.std(pMeanTrim_all)
+    wallShear_mean = np.mean(wallShearStressMeanTrim_all, axis=0)  # (3,)
+    wallShear_std = np.std(wallShearStressMeanTrim_all, axis=0)  # (3,)
+    area_mean = np.mean(area_all)
+    area_std = np.std(area_all)
+
+    logging.info("Statistics computed.")
+
```
Comment (Dibyajyoti-Chakraborty) [inline examples/cfd/external_aerodynamics/moe/preprocessor.py:284]: It might be better if the stats are computed only using the training data. So training is not influenced in any way by validation data.

File: examples/cfd/external_aerodynamics/moe/inference.py [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+        mesh.point_data["fignet_score"] = gate_p[:, 1].cpu().numpy().reshape(-1, 1)
+        mesh.point_data["domino_score"] = gate_p[:, 2].cpu().numpy().reshape(-1, 1)
+
+        # Append predictions to mesh
+        mesh.point_data["pMeanTrimPred_MoE"] = pred_p_denorm.reshape(-1)
+        mesh.point_data["wallShearStressMeanTrimPred_MoE"] = pred_shear_denorm
+
+        # Save mesh
+        out_path = os.path.join(cfg.output_dir, f"{sample['id']}_moe_pred.vtp")
+        mesh.save(out_path)
```
Comment (Dibyajyoti-Chakraborty) [inline examples/cfd/external_aerodynamics/moe/inference.py:209]: Why is the `gate_shear` not saved here ?

File: examples/cfd/external_aerodynamics/moe/conf/config.yaml [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+# │           Model Configuration             │
+# └───────────────────────────────────────────┘
+
+hidden_dim: 128                                            # Hidden dimension of the gating networks
+num_layers: 3                                              # Number of layers in the gating networks
+activation: relu                                           # Activation function (relu, tanh, sigmoid, etc.)
+num_experts: 3                                             # Number of expert models (XMeshGraphNet, FigNet, Domino)
+num_feature_per_expert_pressure: 1                         # Number of features per expert for pressure prediction
+num_feature_per_expert_shear: 3                            # Number of features per expert for shear stress prediction
+use_moe_bias: false                                        # Whether to use bias in the MoE output layer
```
Comment (Dibyajyoti-Chakraborty) [inline examples/cfd/external_aerodynamics/moe/conf/config.yaml:57]: I feel the default setting should be true, given the benefits of the bias term.

Comment (mnabian) [reply examples/cfd/external_aerodynamics/moe/conf/config.yaml:57]: For this experiment, the bias term increased the L-2 errors a bit so I set it to False in my final run.

File: examples/cfd/external_aerodynamics/moe/conf/config.yaml (hunk 1 above)
Comment (mnabian) [reply examples/cfd/external_aerodynamics/moe/conf/config.yaml:66]: I fixed the readme. Batch size is set to 1

File: examples/cfd/external_aerodynamics/moe/inference.py (hunk 3 above)
Comment (mnabian) [reply examples/cfd/external_aerodynamics/moe/inference.py:209]: Good point, added that

File: examples/cfd/external_aerodynamics/moe/preprocessor.py (hunk 2 above)
Comment (mnabian) [reply examples/cfd/external_aerodynamics/moe/preprocessor.py:284]: Ah yeah very good point. Modified the script to exclude validation samples from stats computation.

File: examples/cfd/external_aerodynamics/moe/conf/config.yaml (hunk 4 above)
Comment (Dibyajyoti-Chakraborty) [reply examples/cfd/external_aerodynamics/moe/conf/config.yaml:57]: I see. That makes sense.

File: examples/cfd/external_aerodynamics/moe/README.md [hunk 5]
```diff
@@ -0,0 +1,245 @@
+
+# MoE Gating Network for External Aerodynamics
+
+This recipe contains a **Mixture of Experts (MoE) gating network** implementation
+for combining predictions from multiple aerodynamic models. The system learns to
+optimally weight predictions from different neural network models (DoMINO, FigConvNet,
+and X-MeshGraphNet) to improve overall prediction accuracy for pressure and shear stress
```
Comment (ktangsali) [inline examples/cfd/external_aerodynamics/moe/README.md]: "wall shear stress"

File: examples/cfd/external_aerodynamics/moe/README.md [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+
+$$
+P_{\text{MoE}} = W_1 \cdot P_{\text{DoMINO}} + W_2 \cdot P_{\text{FigConvNet}} + \\
+W_3 \cdot P_{\text{XMGN}} + C_{\text{MoE}}
+$$
+
+Both the weights \( W_1, W_2, W_3 \) and the bias term \( C_{\text{MoE}} \) are
+outputs of the gating network.
+
+#### Architecture Details
```
Comment (ktangsali) [inline examples/cfd/external_aerodynamics/moe/README.md:39]: Just some nit: I think introducing this as a regular MLP and then doing all the MOE operations in the training loop would be better.

File: examples/cfd/external_aerodynamics/moe/README.md [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+```bash
+python preprocessor.py
+```
+
+Given the predictions of the three expert models, this script
+combines predictions from all three expert models,
+extracts surface geometry and field data,
+computes normalization statistics,
+splits data into training and validation sets,
+and saves processed VTP files with all required fields.
```
Comment (ktangsali) [inline examples/cfd/external_aerodynamics/moe/README.md]: nit: `.vtp` - lower case

File: examples/cfd/external_aerodynamics/moe/README.md [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+```
+
+## Model Outputs
+
+### Training Metrics
+
+The training tracks several key metrics:
+
+- **Pressure Loss**: MSE between predicted and true pressure
+- **Shear Loss**: MSE between predicted and true shear stress (WSS-x, WSS-y, WSS-z)
```
Comment (ktangsali) [inline examples/cfd/external_aerodynamics/moe/README.md]: "wall shear stress"

File: examples/cfd/external_aerodynamics/moe/dataset.py [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+import glob
+import numpy as np
+import pyvista as pv
```
Comment (ktangsali) [inline examples/cfd/external_aerodynamics/moe/dataset.py:20]: Since you are using this, might be good to add a requirements.txt

File: examples/cfd/external_aerodynamics/moe/dataset.py [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+        vtp_path = self.files[idx]
+        mesh = pv.read(vtp_path)
+
+        # True fields
+        p = mesh.point_data["pMeanTrim"]
+        shear = mesh.point_data["wallShearStressMeanTrim"]  # shape (N,3)
+        area = mesh.cell_data["Area"]
+        normals = mesh.point_data["Normals"]
+
+        # Predicted fields
```
Comment (ktangsali) [inline examples/cfd/external_aerodynamics/moe/dataset.py:59]: Any specific reason why point data was chosen here? I believe during the benchmarking excercise we decided to use the solutions at the cell centers for the surfaces correct ? Does not matter much, but curious if that was an implementation level detail that was important for training.

File: examples/cfd/external_aerodynamics/moe/README.md [hunk 11]
```diff
@@ -0,0 +1,245 @@
...
+Maximizing entropy helps produce more reliable scores by preventing
+the gating network from over-relying on a single expert, instead
+encouraging balanced contributions from multiple models,
+especially under uncertainty.
+
+In practice, the entropy is computed for each point and averaged
+over the batch. Since entropy is naturally non-negative,
+it is **subtracted** from the total loss (i.e., we maximize it),
+weighted by a coefficient `lambda_entropy`:
+
```
Comment (ktangsali) [inline examples/cfd/external_aerodynamics/moe/README.md:200]: This equation is not being rendered correctly. Please review https://github.com/mnabian/physicsnemo/tree/moe/examples/cfd/external_aerodynamics/moe

File: examples/cfd/external_aerodynamics/moe/README.md [hunk 12]
```diff
@@ -0,0 +1,245 @@
...
+To allow the MoE model to correct for systematic biases—such as all experts
+consistently overpredicting or underpredicting—a bias term \( C_{\text{MoE}} \)
+can be added:
+
+$$
+P_{\text{MoE}} = W_1 \cdot P_{\text{DoMINO}} + W_2 \cdot P_{\text{FigConvNet}} + \\
+W_3 \cdot P_{\text{XMGN}} + C_{\text{MoE}}
+$$
+
+Both the weights \( W_1, W_2, W_3 \) and the bias term \( C_{\text{MoE}} \) are
```
Comment (ktangsali) [inline examples/cfd/external_aerodynamics/moe/README.md]: These terms are not formatted correctly. Please review: https://github.com/mnabian/physicsnemo/tree/moe/examples/cfd/external_aerodynamics/moe

File: examples/cfd/external_aerodynamics/moe/README.md [hunk 13]
```diff
@@ -0,0 +1,245 @@
+
```
Comment (ktangsali) [inline examples/cfd/external_aerodynamics/moe/README.md:1]: Would be nice to add some notes on customizability. For example, what happens if I want to add an expert?

**Review by ktangsali — COMMENTED** [review]
Looks good! Neat result. Added a few comments. Would be nice if we could replace all occurences of "shear-stress" with "wall-shear-stress" for consistency and explainability (or even wss).

File: examples/cfd/external_aerodynamics/moe/README.md (hunk 5 above)
Comment (mnabian) [reply examples/cfd/external_aerodynamics/moe/README.md]: Fixed it

File: examples/cfd/external_aerodynamics/moe/README.md (hunk 6 above)
Comment (mnabian) [reply examples/cfd/external_aerodynamics/moe/README.md:39]: Done

File: examples/cfd/external_aerodynamics/moe/README.md (hunk 7 above)
Comment (mnabian) [reply examples/cfd/external_aerodynamics/moe/README.md]: Fixed

File: examples/cfd/external_aerodynamics/moe/README.md (hunk 8 above)
Comment (mnabian) [reply examples/cfd/external_aerodynamics/moe/README.md]: Fixed

File: examples/cfd/external_aerodynamics/moe/dataset.py (hunk 9 above)
Comment (mnabian) [reply examples/cfd/external_aerodynamics/moe/dataset.py:20]: added

File: examples/cfd/external_aerodynamics/moe/dataset.py (hunk 10 above)
Comment (mnabian) [reply examples/cfd/external_aerodynamics/moe/dataset.py:59]: Hmm no that was not important for training. It can be done with cell data as well. All the models are compared using the point data here, so I don't thing that matters much if we use point or cell data, as long as the comparison is consistent.

File: examples/cfd/external_aerodynamics/moe/README.md (hunk 11 above)
Comment (mnabian) [reply examples/cfd/external_aerodynamics/moe/README.md:200]: Fixed

File: examples/cfd/external_aerodynamics/moe/README.md (hunk 12 above)
Comment (mnabian) [reply examples/cfd/external_aerodynamics/moe/README.md]: Fixed

File: examples/cfd/external_aerodynamics/moe/README.md (hunk 13 above)
Comment (mnabian) [reply examples/cfd/external_aerodynamics/moe/README.md:1]: Added a paragraph about how to add a new expert

**Comment by mnabian** [comment]
Thanks @ktangsali for your comments! All are addressed, please review again.

**Review by ktangsali — APPROVED** [review]
Thanks for addressing the comments. I believe there are still some instances where things like "Shear loss" etc. are used (readme). Not the biggest concern, so I will approve right now to unblock.
